SNT/NativeMemoryManager.cs

```diff
@@ -0,0 +1,133 @@
+        private IntPtr memory;
+        private int length;
+
+        public NativeMemory(IntPtr memory, int length)
+        {
+            if (length < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(length), "Length cannot be negative.");
+            }
+
+            this.memory = memory;
+            this.length = length;
+        }
+
+        /// <summary>
+        /// Destructor for the native memory.
+        /// The Tensor memory lifecycle is managed by Torch. Disposing the memory
+        /// from here will through a "double free or corruption" error.
+        /// <summary>
+        ~NativeMemory()
+        {
+            Dispose(false);
+        }
+
+        /// <summary>
+        /// Whether the memory is disposed or not
+        /// </summary>
+        public bool IsDisposed { get; private set; } = false;
+
+        /// <summary>
+        /// Returns a span wrapping the underlying memory.
+        /// Remember to Unpin the memory once the span is disposed.
+        /// </summary>
+        public override Span<T> GetSpan()
+        {
+            unsafe
+            {
+                return new Span<T>(memory.ToPointer(), length);
+            }
+        }
+
+        /// <summary>
+        /// Returns a handle to the memory that has been pinned and hence its address can be taken.
+        /// </summary>
+        /// <param name="elementIndex">The offset to the element within the memory at which the returned <see cref="MemoryHandle"/> points to. (default = 0)</param>
+        public override MemoryHandle Pin(int elementIndex = 0)
+        {
+            unsafe
+            {
+                Retain();
+                if ((uint)elementIndex > length) throw new ArgumentOutOfRangeException(nameof(elementIndex), "Index out of array bound.");
+                void* pointer = Unsafe.Add<T>((void*)memory, elementIndex);
+                return new MemoryHandle(pointer, default, this);
+            }
+        }
+
+        /// <summary>
+        /// Lets the garbage collector know that the object is free to be moved now.
+        /// </summary>
+        public override void Unpin()
+        {
+            Release();
+        }
+
+        /// <summary>
+        ///   Releases the tensor and its associated data.
+        /// </summary>        
+        public void Dispose()
+        {
+            Dispose(false);
+            GC.SuppressFinalize(this);
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (IsDisposed)
+            {
+                return;
+            }
+
+            if (!Equals(refCount, 0))
+            {
+                throw new InvalidOperationException("Disposing over unmatched Release/Retain");
+            }
+
+            IsDisposed = true;
+
+            if (disposing)
+            {
+                // Typically this would call into a native method appropriate for the platform
+                Marshal.FreeHGlobal(memory);
```

Comment:
I would expect this to PInvoke into torch to tell it to free the memory, or decrement some refcount if it happens to be shared.

Reply:
Thant's a good point. In practice the memory should never be disposed from `NativeMemory` (and in fact disposing is `false`).

Reply:
I guess that raises a different question then: which managed object is responsible for freeing this memory?

Reply:
Some managed object **must** maintain a refcount to the Torch tensor.  When that managed object is disposed or GC'ed it needs to tell Torch that it is done with that Tensor so that Torch knows it is safe to free it at that point.  What's the mechanism that handles this?

Reply:
The refcount is directly managed by Torch, that's why I am forwarding all operations to the Torch layer. Like, in case of clone or resize, Torch maintain the refcount of the storage (memory) so that multiple tensors can point to the same memory and Torch take track of when is safe to free the memory.

Reply:
It looks to me like you're handling this with HType, so that HType is the thing with a finalizer responsible for telling Torch it is OK to free the memory.  In that case then I believe this type should be dealing in terms of HType's as SafeHandles instead of IntPtrs.  That way your NativeMemory objects will keep the SafeHandle's rooted.

Reply:
You are referring to the memory pointer, right? HType and IntPtr are 2 different things here though. HType is a pointer to the tensor (storage) object in Torch, while IntPtr here is the actual pointer to the memory. HType is maintained by the tensor wrapper, not NativeMemory, which instead only surface the native memory. That's also why memory is not cleaned by NativeMemory, it will be freed by the tensor wrapper (by calling the proper free method on the Torch side). Does it make sense?

Reply:
I see what you mean. The workflow is like this:
1) create a tensor
2) retrieve the memory of the tensor by the Buffer property
3) the tensor is destroyed
4) try to do some operation on the memory.
Let me write a test case for this and if this is the case I will write a fix.

Reply:
I fixed this. Now `Native Memory` points to the Torch storage instead of directly to the data pointer. Torch manages the ref count etc.

Reply:
I see.  I'm taking a closer look at these storage classes and they are actually very close to NativeMemory in terms of functionality.  I wonder if you could collapse the two and have the storage classes derive from MemoryManager<T>?